### PR TITLE
login

### PR DESCRIPTION
**Pull request by NguyenHai2310**

### CODE REVIEW

File: app/views/layouts/_header.html.erb [hunk 1]
```diff
@@ -1,5 +1,24 @@
 <header class="navbar navbar-fixed-top navbar-inverse" id="my-header">
   <div class="container">
     <%= link_to t("app_name"), root_path, id: "logo" %>
+    <nav>
+      <ul class="nav navbar-nav navbar-right">
+        <li><%= link_to "Home", root_path %></li>
```
Comment (phuong1492) [inline app/views/layouts/_header.html.erb]: I18n

File: app/views/layouts/_header.html.erb [hunk 2]
```diff
@@ -1,5 +1,24 @@
...
+        <% if user_signed_in? %>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+            <strong><%= current_user.email %></strong><b class="caret"></b></a>
```
Comment (phuong1492) [inline app/views/layouts/_header.html.erb]: ident

File: app/views/users/registrations/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    <%= f.label :current_password %> <i>(we need your current password to confirm your changes)</i><br />
+    <%= f.password_field :current_password, autocomplete: "off" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Update" %>
+  </div>
+<% end %>
+
+<h3>Cancel my account</h3>
```
Comment (phuong1492) [inline app/views/users/registrations/edit.html.erb]: I18n

File: app/views/users/registrations/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+<p>Unhappy? <%= button_to "Cancel my account", registration_path(resource_name), data: { confirm: "Are you sure?" }, method: :delete %></p>
```
Comment (phuong1492) [inline app/views/users/registrations/edit.html.erb]: I18n

File: app/views/users/registrations/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (phuong1492) [inline app/views/users/registrations/edit.html.erb]: I18n

File: app/views/users/registrations/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
+  <%= devise_error_messages! %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+    <div>Currently waiting confirmation for: <%= resource.unconfirmed_email %></div>
```
Comment (phuong1492) [inline app/views/users/registrations/edit.html.erb]: I18n

File: app/views/users/sessions/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<h1><%= t "login" %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
+      <%= f.label :email %><br />
+      <%= f.email_field :email, autofocus: true, class: "form-control" %>
+      <%= f.label :password %><br />
+      <%= f.password_field :password, autocomplete: "off", class: "form-control" %>
+      <% if devise_mapping.rememberable? -%>
+          <div class="field">
```
Comment (phuong1492) [inline app/views/users/sessions/new.html.erb]: ident

File: app/views/users/registrations/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+      </div>
+
+      <div class="field">
+        <%= f.label :age %><br />
+        <%= f.number_field :age, class: "form-control" %>
+      </div>
+
+      <div class="field">
+        <%= f.label :phonenumber %><br />
+        <%= f.number_field :phone_number, class: "form-control" %>
```
Comment (phuong1492) [inline app/views/users/registrations/new.html.erb]: should use f.text_field

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class UsersController < ApplicationController
+
```
Comment (ThuBM) [inline app/controllers/users_controller.rb]: remove this line

File: app/views/users/registrations/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,37 @@
+<h2><%= t "edit" %> <%= resource_name.to_s.humanize %></h2>
+
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
```
Comment (ThuBM) [inline app/views/users/registrations/edit.html.erb]: indent html: {method: :put}

File: app/views/users/registrations/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    <%= f.label :current_password %> <i>(we need your current password to confirm your changes)</i><br />
+    <%= f.password_field :current_password, autocomplete: "off" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Update" %>
+  </div>
+<% end %>
+
+<p><%= t "unhappy" %><%= button_to t("cancel_account"), registration_path(resource_name), data: { confirm: t("you_sure") }, method: :delete %></p>
```
Comment (ThuBM) [inline app/views/users/registrations/edit.html.erb]: indent

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -2,4 +2,10 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.for(:sign_up) << [:name, :age, :address, :job, :phone_number]
```
Comment (daothanhcam) [inline app/controllers/application_controller.rb]: This line have more than 80 characters.

File: app/views/layouts/_header.html.erb [hunk 13]
```diff
@@ -1,5 +1,25 @@
...
+        <li><%= link_to t("home"), root_path %></li>
+        <% if user_signed_in? %>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              <strong><%= current_user.email %></strong><b class="caret"></b>
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to t("header.edit_user"), edit_user_registration_path, class: "navbar-link" %></li>
+              <li class="divider"></li>
+              <li><%= link_to t("logout"), destroy_user_session_path, method: :delete, class: "navbar-link" %></li>
```
Comment (daothanhcam) [inline app/views/layouts/_header.html.erb]: Line too long.

File: app/views/layouts/_header.html.erb [hunk 14]
```diff
@@ -1,5 +1,25 @@
...
+    <nav>
+      <ul class="nav navbar-nav navbar-right">
+        <li><%= link_to t("home"), root_path %></li>
+        <% if user_signed_in? %>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              <strong><%= current_user.email %></strong><b class="caret"></b>
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to t("header.edit_user"), edit_user_registration_path, class: "navbar-link" %></li>
```
Comment (daothanhcam) [inline app/views/layouts/_header.html.erb]: Line too long.

File: app/views/users/registrations/edit.html.erb (hunk 10 above)
Comment (daothanhcam) [inline app/views/users/registrations/edit.html.erb]: Line too long.

File: app/views/users/registrations/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+    <div><%= resource.unconfirmed_email %></div>
+  <% end %>
+
+  <div class="field">
+    <%= f.label :password %> <i>(leave blank if you don't want to change it)</i><br />
```
Comment (daothanhcam) [inline app/views/users/registrations/edit.html.erb]: i18n

File: app/views/users/registrations/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    <%= f.password_field :password, autocomplete: "off" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "off" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :current_password %> <i>(we need your current password to confirm your changes)</i><br />
```
Comment (daothanhcam) [inline app/views/users/registrations/edit.html.erb]: i18n

File: app/views/users/registrations/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        <%= f.text_field :phone_number, class: "form-control" %>
+      </div>
+
+      <div class="field">
+        <%= f.label :address %><br />
+        <%= f.text_field :address, class: "form-control" %><br>
+      </div>
+
+      <div class="actions">
+        <%= f.submit t("register") %>
```
Comment (daothanhcam) [inline app/views/users/registrations/new.html.erb]: Try remove ().

File: app/views/users/sessions/new.html.erb [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
+      <%= f.label :email %><br />
+      <%= f.email_field :email, autofocus: true, class: "form-control" %>
+      <%= f.label :password %><br />
+      <%= f.password_field :password, autocomplete: "off", class: "form-control" %>
+      <% if devise_mapping.rememberable? -%>
+        <div class="field">
+          <%= f.check_box :remember_me %>
+          <%= f.label :remember_me, class: "checkbox inline"%>
```
Comment (daothanhcam) [inline app/views/users/sessions/new.html.erb]: missing a space

File: app/controllers/application_controller.rb (hunk 12 above)
Comment (NguyenHai2310) [inline app/controllers/application_controller.rb]: checked. have 76 character

File: app/views/users/registrations/edit.html.erb [hunk 19]
```diff
@@ -0,0 +1,37 @@
+<h2><%= t "edit" %> <%= resource_name.to_s.humanize %></h2>
+
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name), html:{method: :put}) do |f| %>
```
Comment (khanhkhanh) [inline app/views/users/registrations/edit.html.erb]: Remove the outermost ()

File: app/views/users/registrations/edit.html.erb [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    <%= f.password_field :password_confirmation, autocomplete: "off" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :current_password %></br>
+    <%= f.password_field :current_password, autocomplete: "off" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Update" %>
```
Comment (khanhkhanh) [inline app/views/users/registrations/edit.html.erb]: i18n

File: app/views/users/registrations/new.html.erb [hunk 21]
```diff
@@ -0,0 +1,57 @@
+<div class="row" >
+  <div class="col-md-6 col-md-offset-3" >
+    <h1><%= t "register" %></h1>
+    <%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
```
Comment (khanhkhanh) [inline app/views/users/registrations/new.html.erb]: Remove the outermost ()

File: app/views/users/registrations/new.html.erb [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+      <%= devise_error_messages! %>
+
+      <div class="field">
+        <%= f.label :name %><br />
+        <%= f.text_field :name, autofocus: true, class: "form-control"%>
+      </div>
+
+      <div class="field">
+        <%= f.label :email %><br />
+        <%= f.email_field :email, autofocus: true, class: "form-control" %>
```
Comment (khanhkhanh) [inline app/views/users/registrations/new.html.erb]: There are two fields with "autofocus" attribute. Remove one of them

File: app/views/users/sessions/new.html.erb [hunk 23]
```diff
@@ -0,0 +1,22 @@
+<h1><%= t "login" %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(resource, as: resource_name,
```
Comment (khanhkhanh) [inline app/views/users/sessions/new.html.erb]: Remove the outermost ()

File: app/views/users/shared/_links.html.erb [hunk 24]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
+  <%= link_to "Log in", new_session_path(resource_name) %><br />
```
Comment (khanhkhanh) [inline app/views/users/shared/_links.html.erb]: i18n

File: app/views/users/shared/_links.html.erb [hunk 25]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
```
Comment (khanhkhanh) [inline app/views/users/shared/_links.html.erb]: unlsess

Comment (khanhkhanh) [inline app/views/users/shared/_links.html.erb]: ' => "

File: app/views/users/shared/_links.html.erb [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+  <%= link_to "Log in", new_session_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.registerable? && controller_name != 'registrations' %>
+  <%= link_to "Sign up", new_registration_path(resource_name) %><br />
```
Comment (khanhkhanh) [inline app/views/users/shared/_links.html.erb]: i18n

File: app/views/users/shared/_links.html.erb [hunk 27]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
+  <%= link_to "Log in", new_session_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.registerable? && controller_name != 'registrations' %>
```
Comment (khanhkhanh) [inline app/views/users/shared/_links.html.erb]: ' => "

File: app/views/users/shared/_links.html.erb [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+  <%= link_to "Sign up", new_registration_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations' %>
```
Comment (khanhkhanh) [inline app/views/users/shared/_links.html.erb]: Same

File: app/views/users/registrations/new.html.erb (hunk 21 above)
Comment (NguyenHai2310) [inline app/views/users/registrations/new.html.erb]: can't remove

File: app/views/users/sessions/new.html.erb (hunk 23 above)
Comment (NguyenHai2310) [inline app/views/users/sessions/new.html.erb]: can't remove

File: app/views/layouts/_header.html.erb [hunk 29]
```diff
@@ -1,5 +1,27 @@
...
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              <strong><%= current_user.email %></strong><b class="caret"></b>
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to t("header.edit_user"),
+               edit_user_registration_path, class: "navbar-link" %></li>
+              <li class="divider"></li>
+              <li><%= link_to t("logout"), destroy_user_session_path,
+               method: :delete, class: "navbar-link" %></li>
```
Comment (khanhkhanh) [inline app/views/layouts/_header.html.erb]: Indentation

File: app/views/layouts/_header.html.erb [hunk 30]
```diff
@@ -1,5 +1,27 @@
...
+      <ul class="nav navbar-nav navbar-right">
+        <li><%= link_to t("home"), root_path %></li>
+        <% if user_signed_in? %>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              <strong><%= current_user.email %></strong><b class="caret"></b>
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to t("header.edit_user"),
+               edit_user_registration_path, class: "navbar-link" %></li>
```
Comment (khanhkhanh) [inline app/views/layouts/_header.html.erb]: Indentation

File: app/views/users/sessions/new.html.erb [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+     url: session_path(resource_name)) do |f| %>
+      <%= f.label :email %><br />
+      <%= f.email_field :email, autofocus: true, class: "form-control" %>
+      <%= f.label :password %><br />
+      <%= f.password_field :password, autocomplete: "off",
+       class: "form-control" %>
```
Comment (khanhkhanh) [inline app/views/users/sessions/new.html.erb]: Indentation

File: app/controllers/application_controller.rb [hunk 32]
```diff
@@ -2,4 +2,11 @@ class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.for(:sign_up) <<
+    [:name, :age, :address, :job, :phone_number]
```
Comment (khanhkhanh) [inline app/controllers/application_controller.rb]: Indentation

**Pull request closed without merging**